Replacing tuning decided vectorization into transformMap based one

jerryyin opened the pull request:
**Note: Earliest merge after ROCm 5.4 branch**

This PR includes the following:
 - In Gridwisegemm -> Blockwisegemm lowering, swapped the vectorization decision from tuning parameter derived ones to transformMap based one: `bestVectorization()`
   - Plumb the rest of non-coordinate transformed code just to make them work, eventually will rewrite and get rid of them
 - Removed a significant amount of tuning parameter derive process
   - Get rid of the following data types:  `DerivedParams`, `DerivedOutParams`

I believe part of the reason MIOpen do the vectorization derivation is due to their requirement of cannot have compile failure. Therefore they need to manually derive the parameter as much as possible, otherwise templatized computation may not be able to handle outliers. However, we are more flexible in terms that we can do additional steps of lowering to tell whether certain configs are applicable. This gives me more confidence to eventually remove affix tuning parameter pass.

jerryyin committed:
Replacing tuning decided vectorization into transform based one

jerryyin commented on mlir/lib/Dialect/MIOpen/Tuning/GridwiseGemmParams.cpp:
```diff
@@ -884,48 +434,6 @@ LogicalResult PopulateParamsXDL::getKBlocks(ConvolutionContext &ctx,
-      param.gemmKPerBlock *
-      math_util::integer_least_multiple(param.gemmNPerBlock, max_lds_align);
-
-  ldsSize = (a_block_space + b_block_space) * sizeof(float);
-
-  if (ldsSize > 64 * 1024) {
-    return failure();
-  }
```
Note, previously the assumption was:
 - we assume lds size must be counted as fp32 type
 - The upper bound of the size was statically computed as 64 * 1024

For now I'm bringing the same logic over such that this PR try its best to maintain equivalent coverage. I'd file an additional ticket to document/schedule this such that this is data type and device dependent.

jerryyin commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -613,6 +614,12 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<GridwiseGemmOp> {
     LLVM_DEBUG(llvm::dbgs() << "a_block_space: " << a_block_space << "\n");
     LLVM_DEBUG(llvm::dbgs() << "b_block_space: " << b_block_space << "\n");
     LLVM_DEBUG(llvm::dbgs() << "double_block_space: " << block_space << "\n\n");
+
+    // TODO: adjust for data type and device
+    if (block_space * sizeof(float) > 64 * 1024)
```
Migrated over here.

krzysz00 replied:
Ok, so, deeper question: what is the actual limit here?

krzysz00 replied:
But more immediate note: please file a ticket for that TODO

jerryyin replied:
> But more immediate note: please file a ticket for that TODO

Absolutely.

> Ok, so, deeper question: what is the actual limit here?

If I understand it right, the gfx908 limit on lds should be 64k bytes. Therefore the actual limit should account for both data type and size. I guess the only reason we are still fine is due to:
 - Later arch has larger lds size
 - Among supported types, float is the largest 

and since we are bounded by lower bound, this is okay.

jerryyin replied:
It actually just reminded me that this scheduling can get complicated in a case like: (Note this may not be a problem today but I can see it is as we fuse more aggresively)
 - We have consumed pretty much all lds space in convolution due to the nature of tuning favors larger tiles in general.
 - In a fusion pass, it ended up also require some lds space

In each local scope thing are fine because we didn't exceed upper bound of lds. But it will yield a kernel that over allocate and fail at execution time. 

@sjw36 Does fusion do any LDS allocations? If it does we may need some global flag indicating how much space is left globally.

krzysz00 replied:
Fusion doesn't do LDS allocations

jerryyin commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1064,6 +1075,12 @@ struct GridwiseGemmV2RewritePattern
     LLVM_DEBUG(llvm::dbgs()
                << "total_block_space: " << total_block_space << "\n\n");
+
+    // TODO: adjust for data type and device
+    if (total_block_space * sizeof(float) > 64 * 1024)
```
Migrated over here.

krzysz00 commented on mlir/include/mlir/Dialect/MIOpen/Tuning/GridwiseGemmParams.h:
```diff
@@ -148,45 +130,27 @@ class PopulateParams {
-      DerivedParams &gemmBDerivedParam,
-      DerivedBlockGemmParams &blockGemmDerivedParam,
-      DerivedOutParams &gemmCDerivedParam, uint32_t &gridSize);
+      GemmSize &gemmSize, DerivedBlockGemmParams &blockGemmDerivedParam,
```
Note to future us - see if we can drop the `DerivedBlockGemmParams` into a function that gets called in the lowering process - all those parameters are a static function of block size anyway.

jerryyin replied:
Agreed

krzysz00 replied:
Can you file that ticket?

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -74,7 +74,7 @@ static void computeLoadStoreTypeInfo(OpBuilder &b,
   // In case KPack and vector load is used, and we vector load on GemmK
   // dimension (1), use the last dimension (GemmKPack) instead.
-  if ((loadLength > 1) && (kPack > 1) && (vectorDim == GemmK)) {
+  if ((loadLength > 1) && (kPack > 1) && (vectorDim == 1)) {
```
Would it hurt much to use the enum here?

jerryyin replied:
In the immediate next line there's this enum dimension computation based on size:

> vectorDim = sliceLengths.size() - 1;

I'm not sure I completely comprehend what that does so just left the whole thing to be unit32 type.

krzysz00 replied:
That's handling the kpack case

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -613,6 +614,12 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<GridwiseGemmOp> {
+    // TODO: adjust for data type and device
+    if (block_space * sizeof(float) > 64 * 1024)
+      return failure();
+    else
```
Linter will yell at me for the else there, we could just make this a 
```
return success();
```

More interestingly, both `success()` and `failure()` take boolean arguments.
So you can `return failure(block_space > ...);` and that'll do the right thing

jerryyin replied:
Actually linter is able to tell if there's a return for both if and else statement at the end of the return, so this should be fine.

Agreed though that a last catch all return is a better style. Addressed in f5e59f4

krzysz00 replied:
My copy of clang-tidy whines about
```
if (blah) {
 ...
  return foo;
} else {
  ...
}
```

jerryyin replied:
If else also have a return then it shouldn't... Wired.

krzysz00 replied:
Or maybe if { ... return } else { ... return } is fine, I haven't looked

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1627,7 +1642,18 @@ struct GridwiseGemmV2RewritePattern
     }
 
-    uint32_t blockwiseVectorDimA = matrix_a_source_vector_read_dim;
+    auto dimToIndex = [](GemmDimension dim) {
```
I think this is spelled `static_cast<uint32_t>(dim)`
and if it isn't, I'd rather see this as a `static` function and not a lambda

jerryyin replied:
Yep, agreed that a static_cast is cleaner way to do it. Done in f5e59f4.

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1114,23 +1131,19 @@ struct GridwiseGemmV2RewritePattern
+    GemmDimension vectorTiebreaker =
+        (KPack > 1) ? GemmDimension::K : GemmDimension::MorN;
+    std::tie(matrix_a_source_vector_read_dim, matrix_a_source_data_per_read) =
+        bestVectorization(b, op.a(), aCopyPerThread, vectorTiebreaker);
```
I see one problem here: the xdlops path (if you look in `GemmToGridwise`) still applies a separate `kpack` dimension to the inputs. The code here doesn't account for that and so the K dimension won't be analyzed correctly.

Can you add back in `Value unkpack(Value matrix) {}` and pass the result of that to vectorization inference?

krzysz00 replied:
Or, alternatively, move the function that applies kpack out of `--gemm-to-gridwise` into `--grideise-gemm-to-blockwise` and use it's result anywhere that isn't the vectorization stuff

jerryyin replied:
Nice catch, thank you!

> Can you add back in Value unkpack(Value matrix) {} and pass the result of that to vectorization inference?

I hope I did the right thing in f5e59f4.

> Or, alternatively, move the function that applies kpack out of --gemm-to-gridwise into --grideise-gemm-to-blockwise

I figured that was part of the coordinate transforms stuff so just hold off doing that and will instead do it in next PR.

krzysz00 replied:
Yeah, that's next PR

krzysz00 reviewed (changes requested):
Overall, very nice PR, I've mostly got style nits.

However, there is a big bug here, and that's to do with kPack. We need to bring in the `unkpack()` function that the non-xdlops version didn't need, and the vectorization analysis needs to be passed a G x K x M/N matrix

In minor questions, have you checked some of the cases (`-p -x2` mainly, possibly one of the E2E tests) to see if they haven't lost any vectorization?

Secondly, do we have perf?

jerryyin committed:
Addressing review feedbacks

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/ConvToGemm.cpp:
```diff
@@ -135,15 +135,6 @@ static void maybeSetAttr(StringRef attr, Operation *from, Operation *to) {
   if (maybeValue)
     to->setAttr(attr, maybeValue);
 }
-void affixGemmAttributes(Operation *convOp, Operation *gop, OpBuilder &b) {
```
Is maybeSetAttr() dead code now?

jerryyin replied:
It's now gone in 6da74d5. Nice catch!

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -262,6 +262,22 @@ static TransformingForOp createLdsStoreLoop(OpBuilder &b, Location loc,
+    return matrix;
+  BottomUpTMBuilder unkpack(b, {"g", "k", "d", "kpack"}, shape,
+                            matrix.getLoc());
+  unkpack.passThrough({"g", "m"});
```
... for some reason the comment I left didn't show up here, so yeah, spelling mistake and it's surprising that this doesn't explode.

jerryyin replied:
I can dig into this a bit. It is surprising how this can surprisingly hide another bug as you mentioned [here](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/pull/768#discussion_r974620575).

krzysz00 commented on mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1076,10 +1109,13 @@ struct GridwiseGemmV2RewritePattern
     ArrayRef<int64_t> aShape, bShape, cShape;
-    aShape = op.a().getType().template cast<MemRefType>().getShape();
-    bShape = op.b().getType().template cast<MemRefType>().getShape();
+    aShape = matA.getType().template cast<MemRefType>().getShape();
```
Is this safe? Where else do we use aShape?

jerryyin replied:
Addressed in 2bbf5c6. It's not... I was looking at `aShape` and `bShape` only previously and you reminded me that those will determine the `G, K, M, N` sizes which will be different without kpack. Reverted back to op.a and op.b and wired thing about dimension typo caused regression also went away.

jerryyin committed:
Fixing typo

krzysz00 reviewed (commented):
A few notes, but it's mostly good

Wrt getting more substantial testing, I'm happy with a nightly run by hand and a perf report

krzysz00 commented on mlir/test/mlir-miopen-driver/tuning_xdlops.mlir:
```diff
@@ -2,21 +2,13 @@
 // STEP1: nPerWave
-// STEP1: matrix_c_data_per_copy = 4
-// STEP1: matrix_c_dest_vector_write_dim = 4
-// STEP1: matrix_c_source_vector_read_dim = 2
```
So if we want to keep the spirit of this test in, I'd go to `--gridwise-gemm-to-blockwise` and look for a `threadwise_copy_v2` with length `4`

krzysz00 reviewed (approved):
I leave testing to your discression, unless you want me to poke at it
This code seems quite fine
Approved

jerryyin committed:
Fixing a/b global copy size

jerryyin committed:
Enforce source data per read to follow kpack as upper bound

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1114,23 +1150,28 @@ struct GridwiseGemmV2RewritePattern
+        (KPack > 1) ? GemmDimension::K : GemmDimension::MorN;
+    std::tie(matrix_a_source_vector_read_dim, matrix_a_source_data_per_read) =
+        bestVectorization(b, matA, aCopyPerThread, vectorTiebreaker);
+    if (matrix_a_source_vector_read_dim == GemmDimension::K) {
```
This should be guarded by `kpack > 1`

krzysz00 replied:
(because otherwise you're turning off vectorization in the k dimmension for non-kpack and that's going to kill our perf)

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:
```diff
@@ -1114,23 +1150,28 @@ struct GridwiseGemmV2RewritePattern
+
+    std::tie(matrix_b_source_vector_read_dim, matrix_b_source_data_per_read) =
+        bestVectorization(b, matB, bCopyPerThread, vectorTiebreaker);
+    if (matrix_b_source_vector_read_dim == GemmDimension::K) {
```
And similarly

krzysz00 reviewed (changes requested):
I'm leaving a blocker for the kpack stuff mentioned earlier today so this doesn't land during your vacation

krzysz00 commented:
State of the PR?

krzysz00 committed:
Merge branch 'miopen-dialect' into remove-vec-decision-from-tuning-params

krzysz00 committed:
More precise clamping workarounds for the vectorizer

Because the old vectorization logic makes some assumptions about the
relationship between the vectorization dimensions and kPack that don't
need to be true, it can fail if we don't properly clamp the results of
the vectorizer.

These clamps should be removed entirely when moving to the new
load/store loop structure.

krzysz00 committed:
Merge branch 'miopen-dialect' into remove-vec-decision-from-tuning-params

krzysz00 committed:
Make the buffer_load a global_load to handle wide vectors

Also, fix misc. compiler warnings

krzysz00 commented:
Nightly pass! http://rocmhead.amd.com:8080/blue/organizations/jenkins/MLIR%2Fmlir/detail/PR-768/12/pipeline 

Also, preliminary looks at the perf numbers show an average gain of several flops

krzysz00 committed:
Fix some compiler warnings

krzysz00 committed:
Merge branch 'miopen-dialect' into remove-vec-decision-from-tuning-params

krzysz00 committed:
More clamping to fix division by 0 problem

krzysz00 committed:
Recject padding kernels with kPack > 1 in a more principled way

Currently, we were rejecting padding + kPack > 1 by manually setting
kPack to 1 at the end of the tuning parameter inference. This can
cause problems by caksing (kPerBlock * mPerBlock * kPack) / waveSize
to be zero in gridwise gemm, spooking the vectorizer.

So, instead of that change, we just reject "needing to apply pading"
when there's kPack.

I expect this can go away eventually once we don't have the old
global load math, but that's the future.

krzysz00 committed:
Forgot to clang-format

krzysz00 committed:
Merge branch 'miopen-dialect' into remove-vec-decision-from-tuning-params

jerryyin committed:
Rejecting perf config where block size is too large

jerryyin committed:
Fixing clamp caused derive to zero

krzysz00 reviewed (approved):
Since the weeklies passed, approved!

(It'll be nice to get perf, especially since both this and Manupa's PR are likely to land today, but that's a separate issue)

jerryyin committed:
Merge remote-tracking branch 'origin/miopen-dialect' into remove-vec-decision-from-tuning-params

jerryyin committed:
Fixing clang-format issues
